core/types/src/keys.rs

```diff
@@ -0,0 +1,336 @@
+    _kind: PhantomData<KIND>,
+}
+
+/// Explicit conversion to internal key type for backwards compatibility
+impl<ADDR, KIND, KEY: Default + Zeroize> Key<ADDR, KIND, KEY> {
+    pub fn inner(self) -> KEY {
+        self.key
+    }
+}
+
+/// AsRef to internal key type for backwards compatibility
+impl<ADDR, KIND, KEY: Default + Zeroize> AsRef<KEY> for Key<ADDR, KIND, KEY> {
+    fn as_ref(&self) -> &KEY {
+        &self.key
+    }
+}
+
+/// Create a default key object
+impl<ADDR, KIND, KEY: Default + Zeroize> Default for Key<ADDR, KIND, KEY> {
+    fn default() -> Self {
+        Self {
+            key: KEY::default(),
+            _addr: PhantomData,
+            _kind: PhantomData,
+        }
+    }
+}
+
+// Shared public key methods
+
+impl<ADDR, KIND> Key<ADDR, KIND, RistrettoPublic> {
+    /// Fetch public key bytes in compressed form
+    pub fn to_bytes(&self) -> [u8; 32] {
+        self.key.to_bytes()
+    }
+}
+
+/// Fetch the public key for a private key instance
+impl<ADDR, KIND> From<&Key<ADDR, KIND, RistrettoPrivate>> for Key<ADDR, KIND, RistrettoPublic> {
+    fn from(p: &Key<ADDR, KIND, RistrettoPrivate>) -> Self {
+        Self {
+            key: RistrettoPublic::from(&p.key),
+            _addr: PhantomData,
+            _kind: PhantomData,
+        }
+    }
+}
+
+/// Create a public key from [`RistrettoPublic`] object
+impl<ADDR, KIND> From<RistrettoPublic> for Key<ADDR, KIND, RistrettoPublic> {
+    fn from(p: RistrettoPublic) -> Self {
+        Self {
+            key: p,
+            _addr: PhantomData,
+            _kind: PhantomData,
+        }
+    }
+}
+
+/// Attempt to create a public key from a compressed point, wrapping
+/// [`RistrettoPublic::try_from`]
+impl<ADDR, KIND> TryFrom<&[u8; 32]> for Key<ADDR, KIND, RistrettoPublic> {
+    type Error = KeyError;
+
+    fn try_from(p: &[u8; 32]) -> Result<Self, Self::Error> {
+        let key = RistrettoPublic::try_from(p)?;
+        Ok(Self {
+            key,
+            _addr: PhantomData,
+            _kind: PhantomData,
+        })
+    }
+}
+
+/// Access underlying [`RistrettoPoint`] for public keys using [`AsRef`]
+impl<ADDR, KIND> AsRef<RistrettoPoint> for Key<ADDR, KIND, RistrettoPublic> {
+    fn as_ref(&self) -> &RistrettoPoint {
+        self.key.as_ref()
+    }
+}
+
+/// PartialEq for public key objects
+impl<ADDR, KIND> PartialEq for Key<ADDR, KIND, RistrettoPublic> {
+    fn eq(&self, other: &Self) -> bool {
+        self.key == other.key
+    }
+}
+
+/// PartialEq for backwards compatibility with public key objects
+impl<ADDR, KIND> PartialEq<RistrettoPublic> for Key<ADDR, KIND, RistrettoPublic> {
+    fn eq(&self, other: &RistrettoPublic) -> bool {
+        &self.key == other
+    }
+}
+
+/// PartialEq for backwards compatibility with public key objects
+impl<ADDR, KIND> PartialEq<Key<ADDR, KIND, RistrettoPublic>> for RistrettoPublic {
+    fn eq(&self, other: &Key<ADDR, KIND, RistrettoPublic>) -> bool {
+        self == &other.key
+    }
+}
```

Comment:
Potentially less important for public keys, but is `==` here using `CtEq` or just the stdlib `Eq`?

Reply:
It's `CtEq`.